diff -ur ./src/Makefile ../../../grit/dependencies/lua-5.1.4/src/Makefile
--- ./src/Makefile	2008-01-19 14:37:58.000000000 -0500
+++ ../../../grit/dependencies/lua-5.1.4/src/Makefile	2008-12-20 16:00:20.000000000 -0500
@@ -8,13 +8,13 @@
 PLAT= none
 
 CC= gcc
-CFLAGS= -O2 -Wall $(MYCFLAGS)
+CFLAGS= -O3 -Wall $(MYCFLAGS)
 AR= ar rcu
 RANLIB= ranlib
 RM= rm -f
 LIBS= -lm $(MYLIBS)
 
-MYCFLAGS=
+MYCFLAGS=-DLUA_USE_MKSTEMP
 MYLDFLAGS=
 MYLIBS=
 
