Index: RenderSystems/GL/src/GLSL/src/OgreGLSLLinkProgram.cpp
===================================================================
--- RenderSystems/GL/src/GLSL/src/OgreGLSLLinkProgram.cpp	(revision 9227)
+++ RenderSystems/GL/src/GLSL/src/OgreGLSLLinkProgram.cpp	(working copy)
@@ -174,8 +174,13 @@
 					if (pos != String::npos)
 					{
 						String::size_type startpos = vpSource.find("attribute", pos-20);
-						if (startpos != String::npos && startpos < pos)
+						if (startpos != String::npos && startpos < pos) {
 							glBindAttribLocationARB(mGLHandle, a.attrib, a.name.c_str());
+						} else {
+							startpos = vpSource.find("in", pos-20);
+							if (startpos != String::npos && startpos < pos)
+								glBindAttribLocationARB(mGLHandle, a.attrib, a.name.c_str());
+						}
 					}
 				}
 			}
