Index: OgreMain/include/OgreResource.h
===================================================================
--- OgreMain/include/OgreResource.h	(revision 9577)
+++ OgreMain/include/OgreResource.h	(working copy)
@@ -161,7 +161,7 @@
 		/// State count, the number of times this resource has changed state
 		size_t mStateCount;
 
-		typedef list<Listener*>::type ListenerList;
+		typedef set<Listener*>::type ListenerList;
 		ListenerList mListenerList;
 		OGRE_MUTEX(mListenerListMutex)
 
Index: OgreMain/src/OgreResource.cpp
===================================================================
--- OgreMain/src/OgreResource.cpp	(revision 9577)
+++ OgreMain/src/OgreResource.cpp	(working copy)
@@ -344,14 +344,14 @@
 	void Resource::addListener(Resource::Listener* lis)
 	{
 		OGRE_LOCK_MUTEX(mListenerListMutex)
-		mListenerList.push_back(lis);
+		mListenerList.insert(lis);
 	}
 	//-----------------------------------------------------------------------
 	void Resource::removeListener(Resource::Listener* lis)
 	{
 		// O(n) but not called very often
 		OGRE_LOCK_MUTEX(mListenerListMutex)
-		mListenerList.remove(lis);
+		mListenerList.erase(lis);
 	}
 	//-----------------------------------------------------------------------
 	void Resource::_fireLoadingComplete(bool wasBackgroundLoaded)
Index: RenderSystems/Direct3D9/src/OgreD3D9RenderWindow.cpp
===================================================================
--- RenderSystems/Direct3D9/src/OgreD3D9RenderWindow.cpp	(revision 9577)
+++ RenderSystems/Direct3D9/src/OgreD3D9RenderWindow.cpp	(working copy)
@@ -379,10 +379,11 @@
 				}
 				else
 				{
-					SetWindowPos(mHWnd, HWND_TOPMOST, mLeft, mTop, width, height, SWP_NOACTIVATE);
-					//MoveWindow(mHWnd, mLeft, mTop, mWidth, mHeight, FALSE);
 					SetWindowLong(mHWnd, GWL_STYLE, dwStyle);
 					SetWindowPos(mHWnd, 0, 0,0, 0,0, SWP_NOACTIVATE | SWP_NOMOVE | SWP_NOSIZE | SWP_NOZORDER);
+
+                                        SetWindowPos(mHWnd, HWND_TOPMOST, mLeft, mTop, width, height, SWP_NOACTIVATE);
+					//MoveWindow(mHWnd, mLeft, mTop, mWidth, mHeight, FALSE);
 				}
 			}
 			else
@@ -399,7 +400,7 @@
 				SetWindowLong(mHWnd, GWL_STYLE, dwStyle);
 				SetWindowPos(mHWnd, HWND_NOTOPMOST, 0, 0, winWidth, winHeight,
 					SWP_DRAWFRAME | SWP_FRAMECHANGED | SWP_NOMOVE | SWP_NOACTIVATE);
-				// Note that we also set the position in the restoreLostDevice method
+				// Note that we also set the position in the restoreLostDevice methods
 				// via _finishSwitchingFullScreen
 			}
 						
